File: src/Monads/Maybe.js
const isNullOrUndefined = x => x === null || x === undefined

const Maybe = x => (isNullOrUndefined(x) ? Maybe.Nothing(x) : Maybe.Just(x))

Maybe.Just = x => ({
  // map :: Functor f => f a ~> (a -> b) -> f b
  map: f => Maybe.Just(f(x)),
  // flatMap :: Functor f => f a ~> (a -> f b) -> f b
  flatMap: f => f(x),
  // flatten :: Functor f => f (f a) ~> f a
  flatten: () => x,
  // inspect :: Just j => j a ~> () -> String
  inspect: () => `Just(${x})`,
})

Maybe.Nothing = x => ({
  // map :: Functor f => f a ~> (a -> b) -> f a
  map: _f => Maybe.Nothing(x),
  // flatMap :: Functor f => f a ~> (a -> f b) -> f a
  flatMap: _f => Maybe.Nothing(x),
  // flatten :: Functor f => f (f a) ~> f a
  flatten: () => x,
  // inspect :: Nothing n => n a ~> () -> String
  inspect: () => `Nothing(${x})`,
})

export default Maybe

File: src/Monads/Maybe.hs
import Prelude hiding (head, Maybe (..))

data Maybe a = Nothing | Just a deriving (Show)

instance Functor Maybe where
  fmap f (Just x) = Just (f x)
  fmap _ Nothing = Nothing

instance Applicative Maybe where
  pure x = Just x
  (Just f) <*> (Just x) = Just (f x)
  Nothing <*> _ = Nothing
  _ <*> Nothing = Nothing

instance Monad Maybe where 
  Just x >>= f = f x
  Nothing >>= _ = Nothing
  return = pure

head :: [a] -> Maybe a
head (x : _) = Just x
head [] = Nothing

double :: Int -> Int
double x = x * 2

main :: IO ()
main = do 
  print (double . double <$> head [1, 2, 3])
  print (double . double <$> head [])